docs/reference-guides/best-practices/rancher-server/tuning-and-best-practices-for-rancher-at-scale.md

```diff
@@ -21,6 +21,12 @@ This guide describes the best practices and tuning approaches to scale Rancher s
 
 When scaling up Rancher, one typical bottleneck is resource growth in the upstream (local) Kubernetes cluster. The upstream cluster contains information for all downstream clusters. Many operations that apply to downstream clusters create new objects in the upstream cluster and require computation from handlers running in the upstream cluster.
 
+### Avoiding third-party applications
+
+Running workloads or microservices in in the upstream cluster other than Rancher can cause performance issues, especially if the application makes heavy use of CRDs or otherwise creates significant load on Kubernetes components such as `etcd` or `kubeapiserver`.
+
+Among the applications that were found to interfere significantly with Rancher performance include [CrossPlane](https://www.crossplane.io/), [Argo CD](https://argoproj.github.io/cd/) and [Flux](https://fluxcd.io/).
```

Comment:
@btat I'm a little leery of going beyond "apps that create significant load" to naming and linking to specific examples. What do you think?

Reply:
> @btat I'm a little leery of going beyond "apps that create significant load" to naming and linking to specific examples. What do you think?
 
Agreed. Since the recommendation is a blanket statement to avoid any third-party applications it seems fine to not include specific examples.

Reply:
I agree it's a double edged sword - on one hand it might seem a bit contradictory, on the other hand we are seeing this pattern repeating and would like to be more incisive on the biggest known troublemakers.

@pgonin, your opinion?

Reply:
I think this section sends the wrong message to customers (and potential customers). 

 "avoid third party applications" is saying don't run anything else in the Kubernetes cluster apart from Rancher Manager. Should customers avoid CNIs, cloud controller managers, observability etc, all of which can be classed as 3rd party. And thinking about our roadmap should customers not install cluster api providers and related controllers Azure Service Operator?

Also, calling out specific applications like this doesn't seem right. What criteria are we selecting these based on? What other factors where at play in the cluster? Are we going to keep updating the list? How do these specific examples cause performance issues (it can't be based purely on CRDs as Crossplane and Flux are completely different in this respect).

Reply:
> Should customers avoid CNIs, cloud controller managers, observability etc, all of which can be classed as 3rd party. And thinking about our roadmap should customers not install cluster api providers and related controllers Azure Service Operator?

I think you are right with this objection, my wording here is very broad and should be qualified better.

I agree that CNIs, cloud controller managers, observability/monitoring tools, Cluster API components are all acceptable. I admit I am not very knowledgeable of the Kubernetes ecosystem at large, so if you have other ideas about such "categories" that should be whitelisted please let me know.

> Also, calling out specific applications like this doesn't seem right. What criteria are we selecting these based on? What other factors where at play in the cluster? Are we going to keep updating the list? How do these specific examples cause performance issues (it can't be based purely on CRDs as Crossplane and Flux are completely different in this respect)?

The criterion here is: we have seen customer escalations around Rancher performance where the culprit was ultimately interference from those apps. In the specific case of Crossplane which I was involved in, it appeared due to the high number of CRDs installed. In the specific case of ArgoCD, we have seen latency spikes from the Kubernetes API which disappeared after Argo was shut down, and I was not able to further track down the cause in ArgoCD's internals. The Flux case was just reported to me, I do not have details at this time. Yes, it would be my intention to keep the list updated to be extra clear with customers and users they should steer away from known problematic cases at least.

In general, the intent is to discourage the installation of any "_end-user applications_" other than Rancher in the Rancher cluster, as - especially in high scale cases - it can be demanding on resources and stressful to Kubernetes components. We want to recommend a dedicated cluster just for Rancher (and the bare minimum around it). Moreover, applications implementing functionality which is similar to the one provided by Rancher and its ecosystem (eg. Fleet) is at risk of conflict - performance and otherwise. I acknowledge my definition of "_end-user applications_" is vague - the point is we want to avoid the situation where a complicated performance problem ultimately boils down to, picking an extreme example, a very busy Wordpress running side by side to Rancher.

Do you have ideas for a better wording?